tools/integration_tests/concurrent_operations/high_cpu_concurrent_listing_test.go

```diff
@@ -0,0 +1,522 @@
+			}
+		}()
+	}
+
+	// Wait for goroutines or timeout
+	done := make(chan bool, 1)
+	go func() {
+		wg.Wait()
+		done <- true
+	}()
+
+	select {
+	case <-done:
+		// Success: All Readdir operations finished before timeout
+	case <-time.After(timeout):
+		assert.FailNow(t, "Possible deadlock or race condition detected during concurrent Readdir calls")
+	}
+}
+
+// Test_RecursiveListingAndFileRead tests for potential deadlocks or race conditions
+// when multiple goroutines performs recursive listing.
+func (s *highCpuConcurrentListingTest) Test_RecursiveListingAndFileRead(t *testing.T) {
+	if runtime.NumCPU() < requiredCPUCoresToRunThisTest {
+		t.SkipNow()
+	}
+	t.Parallel() // Mark the test parallelizable.
+	testCaseDir := "Test_RecursiveListingAndFileRead"
+	createDirectoryStructureForTestCaseParallel(t, testCaseDir)
+	targetDir := path.Join(testDirPath, testCaseDir, "explicitDir")
+	var wg sync.WaitGroup
+	availableCPU := runtime.NumCPU() / 2 // Keep half for gcsfuse process.
+	t.Logf("Testing with %d go-routine: ", availableCPU)
+	goRoutineCountPerOperation := availableCPU / 2 // Use all cpus for recursive listing.
+	timeout := 400 * time.Second
+
+	// Create multiple go routines to listing concurrently.
+	for r := 0; r < goRoutineCountPerOperation; r++ {
+		wg.Add(2)
+		// Repeatedly do recursive listing.
+		go func() {
+			defer wg.Done()
+			for j := 0; j < iterationsForMediumOperations; j++ {
+				listDirectoryRecursivelyWithCmd(t, targetDir)
+			}
+		}()
+
+		// Repeatedly file read.
+		go func() {
+			defer wg.Done()
+			for j := 0; j < iterationsForMediumOperations; j++ {
+				data, err := os.ReadFile(path.Join(targetDir, "level1/file1.txt"))
+				assert.Nil(t, err)
+				assert.True(t, bytes.Equal(data, []byte("test_content")))
+			}
+		}()
+	}
+
+	// Wait for goroutines or timeout
+	done := make(chan bool, 1)
+	go func() {
+		wg.Wait()
+		done <- true
+	}()
+
+	select {
+	case <-done:
+		// Success: All Readdir operations finished before timeout
+	case <-time.After(timeout):
+		assert.FailNow(t, "Possible deadlock or race condition detected during concurrent Readdir calls")
+	}
+}
+
+// Test_AllReadOperationsTogether tests for potential deadlocks or race conditions
+// when multiple goroutines performs recursive listing, openDir, stat operations with
+// repetitions.
+func (s *highCpuConcurrentListingTest) Test_AllReadOperationsTogether(t *testing.T) {
+	if runtime.NumCPU() < requiredCPUCoresToRunThisTest {
+		t.SkipNow()
+	}
+	t.Parallel() // Mark the test parallelizable.
+	testCaseDir := "Test_AllReadOperationsTogether"
+	createDirectoryStructureForTestCaseParallel(t, testCaseDir)
+	targetDir := path.Join(testDirPath, testCaseDir, "explicitDir")
+	var wg sync.WaitGroup
+	availableCPU := runtime.NumCPU() / 2 // Keep half for gcsfuse process.
+	t.Logf("Testing with %d go-routine: ", availableCPU)
+	goRoutineCountPerOperation := availableCPU / 4 // Divide among three read-only operations.
```

Comment:
The comment says that we are dividing among three operations but we are dividing by 4. Why?